source/fundamentals/gridfs.txt

```diff
@@ -0,0 +1,387 @@
+   :start-after: start uploadOptions
+   :end-before: end uploadOptions
+
+See the :java-docs:`GridFSUploadOptions <apidocs/mongodb-driver-core/com/mongodb/client/gridfs/model/GridFSUploadOptions.html>`
+API documentation for more information.
+
+Upload a File Using an Input Stream
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This section shows you how to upload a file to a GridFS bucket using an input
+stream. The code example below shows how you can use a ``FileInputStream`` to
+read data from a file in your filesystem and upload it to GridFS by performing
+the following operations:
+
+- Read from the filesystem using a ``FileInputStream``.
+- Set the chunk size using ``GridFSUploadOptions``.
+- Set a custom metadata field called ``type`` to the value "zip archive".
+- Upload the file, specifying the file name as "myProject.zip".
+
+.. literalinclude:: /includes/fundamentals/code-snippets/GridFSOperations.java
+   :language: java
+   :dedent:
+   :start-after: start uploadFromInputStream
+   :end-before: end uploadFromInputStream
+
+This code example prints the file id of the uploaded file after it is
+successfully saved in GridFS.
+
+For more information, see the API documentation on
+:java-docs:`uploadFromStream() <apidocs/mongodb-driver-sync/com/mongodb/client/gridfs/GridFSBucket.html#uploadFromStream(java.lang.String,java.io.InputStream,com.mongodb.client.gridfs.model.GridFSUploadOptions)>`.
+
+Upload a File Using an Output Stream
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This section shows you how to upload a file to a GridFS bucket by writing to
+an output stream. The code example below shows how you can write to a
+``GridFSUploadStream`` to send data to GridFS by performing the following
+operations:
+
+- Read a file from the filesystem into a byte array.
+- Set the chunk size using ``GridFSUploadOptions``.
+- Set a custom metadata field called ``type`` to the value "zip archive".
+- Write the bytes to a ``GridFSUploadStream``. The stream reads data into
+  a buffer until it reaches the limit specified in the ``chunkSize``
+  setting, and inserts it as a new chunk in the ``chunks`` collection.
+
+.. literalinclude:: /includes/fundamentals/code-snippets/GridFSOperations.java
+   :language: java
+   :dedent:
+   :start-after: start uploadFromOutputStream
+   :end-before: end uploadFromOutputStream
+
+This code example prints the file id of the uploaded file after it is
+successfully saved in GridFS.
+
+For more information, see the API documentation on
+:java-docs:`GridFSUploadStream <apidocs/mongodb-driver-sync/com/mongodb/client/gridfs/GridFSUploadStream.html>`.
+
+.. _gridfs-retrieve-file-info:
+
+Retrieve File Information
+-------------------------
+
+In this section, you can learn how to retrieve file metadata stored in the
+``files`` collection of the GridFS bucket. The metadata includes information
+about the file it refers to including:
+
+- Id of the file
+- Name of the file
+- Length/size of the file
+- Upload date and time
```

Comment:
nit: since these aren't full sentence bullet points and "Id" looks *very* strange to me with a capital I... would it make sense to use lowercase letters at the beginning of these instead?

Reply:
Our style guide mentions these should be capitalized: https://docs.mongodb.com/meta/style-guide/style/lists/#all-list-items-are-fragments

Reply:
seeing "Id" with a capital I is so jarring to me that I would lean towards adding the article, but use your judgment here -- it's a very subjective call.